In [2]:
import pandas as pd
import numpy as np

## Обзор данных

Для считывания данных используется функция `pd.read_csv`. У этой функции много аргументов (см. документацию), первый из которых - название считываемого файла с данными, второй - sep - разделитель, по умолчанию это запятая. Данные в csv файлах (обычно, но не всегда) записаны в виде "таблицы". Одна запись - одна строка, данные столбцов разделяются запятыми.

Для работы с файлом его, как и обычный текстовый файл (впрочем, это и есть обычный текстовый файл!), удобнее положить в ту же директорию, где лежит ноутбук. Затем вы можете прямо через jupyter открыть файл `csv` и увидеть, как он выглядит.

In [7]:
f = pd.read_csv("telecom_churn.csv")

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

Посмотреть начало и конец таблицы (по умолчанию 5 строк, но это настраиваемый параметр) - функции `head` и `tail`. Описание данных - `describe`, `info`. Размер таблицы - `shape`, названия колонок - `columns`. Обратите внимание, что `columns` возвращает итерируемый объект.

In [ ]:
f.head()

In [ ]:
f.head(7)

In [ ]:
f.tail()

In [ ]:
f.tail(7)

In [ ]:
f.tail(100)

In [8]:
f.shape

(3333, 20)

In [9]:
f.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [10]:
f.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [11]:
f.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Небольшое, но важное отступление. Глубоко понимать это пока не обязательно. Обратите внимание, что функция `print` возвращает строковое представление объекта и она не необходима, чтобы вывести объект. 2 ячейки сработают одинаково:

In [12]:
a = 1
print(a)

1


In [13]:
a = 1
a

1

При этом для некоторых объектов (для которых стандартные функции `__str__` и `__repr__` возвращают разные значения) два способа вывода сработают несколько по-разному:

In [14]:
a = np.array([1.0, 2.0])
print(a)

[1. 2.]


In [15]:
a = np.array([1.0, 2.0])
a

array([1., 2.])

Такая же ситуация с pandas dataframe'ами. Без print получится красивее, чем с ним, но фактически это одно и то же:

In [16]:
print(f)

     State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                   

In [17]:
f

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

2 варианта синтаксиса для просмотра содержимого столбца:

In [18]:
f.State

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

In [19]:
f['State']

0       KS
1       OH
2       NJ
3       OH
4       OK
        ..
3328    AZ
3329    WV
3330    RI
3331    CT
3332    TN
Name: State, Length: 3333, dtype: object

Обратите внимание, что таблица и столбцы имеют разный тип:

In [20]:
type(f), type(f.State), type(f['State'])

(pandas.core.frame.DataFrame,
 pandas.core.series.Series,
 pandas.core.series.Series)

Тем не менее, мы можем сделать именно датафрейм из одного столбца следующим образом:

In [ ]:
f[['State']]

In [ ]:
type(f[['State']]), type(f['State'])

Это способ обращения к "срезам" таблицы по названиям столбцов. Еще пример:

In [21]:
f[['State', 'Churn']]

State  Churn
0       KS  False
1       OH  False
2       NJ  False
3       OH  False
4       OK  False
...    ...    ...
3328    AZ  False
3329    WV  False
3330    RI  False
3331    CT  False
3332    TN  False

[3333 rows x 2 columns]

Просмотр уникальных значений в столбце и их числа:

In [22]:
f.State.unique()

array(['KS', 'OH', 'NJ', 'OK', 'AL', 'MA', 'MO', 'LA', 'WV', 'IN', 'RI',
       'IA', 'MT', 'NY', 'ID', 'VT', 'VA', 'TX', 'FL', 'CO', 'AZ', 'SC',
       'NE', 'WY', 'HI', 'IL', 'NH', 'GA', 'AK', 'MD', 'AR', 'WI', 'OR',
       'MI', 'DE', 'UT', 'CA', 'MN', 'SD', 'NC', 'WA', 'NM', 'NV', 'DC',
       'KY', 'ME', 'MS', 'TN', 'PA', 'CT', 'ND'], dtype=object)

In [23]:
f.State.nunique()

51

Обращение к элементу ообъекта `pd.Series`:

In [58]:
f.State[0]

'KS'

Полный набор функций, применимых к столбцу можно посмотреть следующим образом. Здесь используется стандартная функция `__dir__()`, которая выводит список атрибутов и методов объекта.

In [ ]:
f.State.__dir__()

С `pd.DataFrame` привычный способ обращения по индексам не сработает:

In [ ]:
f[0][0]

Основной способ - использование `loc` и `iloc`. `loc` дает индексацию по "настоящим" индексам - названиям столбцов и названиям строк. В нашем случае, названия строк - целые числа, как обычные индексы. `iloc` позволяет осуществлять стандартную численную индексацию, как для `numpy` массивов.

In [25]:
f.iloc[0] # первая строка

State                        KS
Account length              128
Area code                   415
International plan           No
Voice mail plan             Yes
Number vmail messages        25
Total day minutes         265.1
Total day calls             110
Total day charge          45.07
Total eve minutes         197.4
Total eve calls              99
Total eve charge          16.78
Total night minutes       244.7
Total night calls            91
Total night charge        11.01
Total intl minutes         10.0
Total intl calls              3
Total intl charge           2.7
Customer service calls        1
Churn                     False
Name: 0, dtype: object

In [26]:
f.iloc[0][0]

'KS'

In [33]:
f.iloc[0, 0]

'KS'

In [ ]:
f.iloc[[0, 0]]

In [ ]:
f.loc[0, 0]

In [27]:
f.iloc[[1, 2, 5, 1]]

State  Account length  Area code International plan Voice mail plan  \
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
5    AL             118        510                Yes              No   
1    OH             107        415                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
1                     26              161.6              123   
2                      0              243.4              114   
5                      0              223.4               98   
1                     26              161.6              123   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
5             37.98              220.6              101             18.75   
1             27.47              195.5              103             16.62   

   Total night minutes  Total night calls  Total night charge  \
1                254.4                103               11.45   
2                162.6                104                7.32   
5                203.9                118                9.18   
1                254.4                103               11.45   

   Total intl minutes  Total intl calls  Total intl charge  \
1                13.7                 3               3.70   
2                12.2                 5               3.29   
5                 6.3                 6               1.70   
1                13.7                 3               3.70   

   Customer service calls  Churn  
1                       1  False  
2                       0  False  
5                       0  False  
1                       1  False

In [ ]:
f.loc[[1, 2, 5, 1]]

In [28]:
f.loc[[1, 2, 5, 1], ["State", "Account length"]]

State  Account length
1    OH             107
2    NJ             137
5    AL             118
1    OH             107

In [29]:
f.iloc[[1, 2, 5, 1], [0, 1]]

State  Account length
1    OH             107
2    NJ             137
5    AL             118
1    OH             107

Следующая ячейка вернет ошибку:

In [31]:
f.iloc[[1, 2, 5, 1], ["State", "Account length"]]

IndexError: .iloc requires numeric indexers, got ['State' 'Account length']

In [30]:
f.iloc[:5, :4]

State  Account length  Area code International plan
0    KS             128        415                 No
1    OH             107        415                 No
2    NJ             137        415                 No
3    OH              84        408                Yes
4    OK              75        415                Yes

In [ ]:
f.loc[:, ["State", "Account length"]]

In [ ]:
f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

Обратите внимание, что `loc` включает и строки соответствующие и начальному, и конечному индексам.

Сохраним этот датафрейм в новую переменную и рассмотрим отличие loc и iloc в работе по целочисленным индексам строк

In [37]:
new_f = f.loc[4:9, ["State", "Account length", "Area code", "International plan"]]

In [38]:
new_f.iloc[:5]

State  Account length  Area code International plan
4    OK              75        415                Yes
5    AL             118        510                Yes
6    MA             121        510                 No
7    MO             147        415                Yes
8    LA             117        408                 No

In [39]:
new_f.loc[:5]

State  Account length  Area code International plan
4    OK              75        415                Yes
5    AL             118        510                Yes

**Упражнение 1.** Объясните различие поведения `loc` и `iloc` для `new_f`.

In [45]:
#loc включает оба конца среза, а iloc полу интервал

#### Создание датафрейма и присваивание по индексам

In [82]:
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]}, index = ["first", "second", "third"])
df

A  B
first   1  4
second  2  5
third   3  6

In [83]:
df.iloc[0, 0] = 100
df

A  B
first   100  4
second    2  5
third     3  6

In [84]:
df.loc["second", "B"] = 31
df

A   B
first   100   4
second    2  31
third     3   6

In [85]:
df.iloc[2] = 5
df

A   B
first   100   4
second    2  31
third     5   5

In [86]:
df.loc["B"] = 0
df

A   B
first   100   4
second    2  31
third     5   5
B         0   0

In [87]:
df.loc[:, "B"] = 0
df

A  B
first   100  0
second    2  0
third     5  0
B         0  0

In [88]:
df = pd.DataFrame({"A": [1, 2, 3, 4], "B": [5, 6, 7, 8]})
df

A  B
0  1  5
1  2  6
2  3  7
3  4  8

In [ ]:
df.iloc[1:2, :] = -1
df

In [ ]:
df.loc[1:2, :] = -2
df

Еще раз убеждаемся, что `loc` включает оба конца "среза", а `iloc` - полуинтервал, как например обычные срезы.

#### Создание нового столбца

In [89]:
df["C"] = 1
df

A  B  C
0  1  5  1
1  2  6  1
2  3  7  1
3  4  8  1

In [92]:
df["D"] = df["A"] + df["B"]
df

A  B  C   D
0  1  5  1   6
1  2  6  1   8
2  3  7  1  10
3  4  8  1  12

In [93]:
df["E"] = df["D"].apply(lambda x: x ** 2)
df

A  B  C   D    E
0  1  5  1   6   36
1  2  6  1   8   64
2  3  7  1  10  100
3  4  8  1  12  144

In [94]:
df["F"] = df[["A", "C", "E"]].apply(lambda x: x.sum(), axis=1)
df

A  B  C   D    E    F
0  1  5  1   6   36   38
1  2  6  1   8   64   67
2  3  7  1  10  100  104
3  4  8  1  12  144  149

**Упражнение 2.** Объясните последний результат.

In [95]:
#Столбец Ф = суммированию А, С , Е по горизонтальной оси

## Срезы с условиями и аггрегация

In [ ]:
f[f.Churn == True]

In [ ]:
f[f["Area code"] == 408]

In [ ]:
f[f["Area code"] == 408].loc[3]

In [ ]:
f[f["Area code"] == 408].loc[3] == f[f["Area code"] == 408].iloc[0]

In [ ]:
f[f.Churn == True]

In [ ]:
f[f.Churn == True].mean()

In [ ]:
f[f.Churn == True].sum()

Выборка по нескольким условиям требует скобок, т.к. операции `&`, `|`, `~` (и, или, не) имеют более высокий приоритет, чем `==`, `>=`, `<=` и арифметические:

In [ ]:
f[(f.Churn == True) & (f["Area code"] == 408)]

Можем посмотреть средние показатели лояльных и нелояльных клиентов (столбец "Churn" показывает, ушел клиент от нас как от оператора связи или нет).

In [ ]:
loyal = f[f.Churn == False].mean()
unloyal = f[f.Churn == True].mean()
pd.DataFrame({"loyal": loyal, "unloyal": unloyal})

`groupby` позволяет делать группировку по уникальным значениям столбца, которые будут играть роль индекса в полученном датафрейме. При этом также нужна функция агрегации (`mean`, `sum`, `max` и прочие), чтобы отобразить полученный объект.

In [ ]:
f.groupby("Churn")

In [67]:
f.groupby("Churn").mean()

Account length   Area code  Number vmail messages  Total day minutes  \
Churn                                                                         
False      100.793684  437.074737               8.604561         175.175754   
True       102.664596  437.817805               5.115942         206.914079   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
Churn                                                                          
False       100.283158         29.780421         199.043298       100.038596   
True        101.335404         35.175921         212.410145       100.561077   

       Total eve charge  Total night minutes  Total night calls  \
Churn                                                             
False         16.918909           200.133193         100.058246   
True          18.054969           205.231677         100.399586   

       Total night charge  Total intl minutes  Total intl calls  \
Churn                                                             
False            9.006074           10.158877          4.532982   
True             9.235528           10.700000          4.163561   

       Total intl charge  Customer service calls  
Churn                                             
False           2.743404                1.449825  
True            2.889545                2.229814

Здесь мы сделали агрегацию всех столбцов функцией `mean`. Часто нам нужна агрегация по конкретному или конкретным столбцам. Посмотрим, например, в каком штате больше всего разговаривают по телефону:

In [63]:
f.groupby(["State"]).agg({"Total day minutes": "mean"})

Total day minutes
State                   
AK            178.384615
AL            186.010000
AR            176.116364
AZ            171.604688
CA            183.564706
CO            178.712121
CT            175.140541
DC            171.379630
DE            174.583607
FL            179.533333
GA            185.025926
HI            175.962264
IA            177.613636
ID            178.619178
IL            173.591379
IN            196.525352
KS            191.555714
KY            173.754237
LA            178.376471
MA            180.103077
MD            197.228571
ME            185.262903
MI            180.593151
MN            183.354762
MO            170.506349
MS            177.929231
MT            174.007353
NC            185.145588
ND            187.338710
NE            177.465574
NH            177.328571
NJ            196.225000
NM            171.429032
NV            176.425758
NY            175.114458
OH            183.274359
OK            179.909836
OR            176.246154
PA            188.375556
RI            167.478462
SC            166.441667
SD            189.690000
TN            175.771698
TX            181.516667
UT            183.569444
VA            177.244156
VT            182.031507
WA            178.742424
WI            179.130769
WV            173.950943
WY            180.170130

Максимальное значение времени:

In [53]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).max()

Total day minutes    197.228571
dtype: float64

Чтобы найти сам штат, можем отсортировать данные:

In [ ]:
f.groupby(["State"]).agg({"Total day minutes": "mean"}).sort_values(by="Total day minutes", ascending = False)

**Пример.** Как распределен отток клиентов по штатам? В каких штатах он больше среднего?

In [66]:
f.groupby(["State"]).agg({"Churn": "mean"})

Churn
State          
AK     0.057692
AL     0.100000
AR     0.200000
AZ     0.062500
CA     0.264706
CO     0.136364
CT     0.162162
DC     0.092593
DE     0.147541
FL     0.126984
GA     0.148148
HI     0.056604
IA     0.068182
ID     0.123288
IL     0.086207
IN     0.126761
KS     0.185714
KY     0.135593
LA     0.078431
MA     0.169231
MD     0.242857
ME     0.209677
MI     0.219178
MN     0.178571
MO     0.111111
MS     0.215385
MT     0.205882
NC     0.161765
ND     0.096774
NE     0.081967
NH     0.160714
NJ     0.264706
NM     0.096774
NV     0.212121
NY     0.180723
OH     0.128205
OK     0.147541
OR     0.141026
PA     0.177778
RI     0.092308
SC     0.233333
SD     0.133333
TN     0.094340
TX     0.250000
UT     0.138889
VA     0.064935
VT     0.109589
WA     0.212121
WI     0.089744
WV     0.094340
WY     0.116883

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).loc["LA"]

In [55]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()

Churn    0.144927
dtype: float64

In [ ]:
f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]

In [115]:
m = f.groupby(["State"]).agg({"Churn": "mean"}).mean()[0]
new_dataframe = f.groupby(["State"]).agg({"Churn": "mean"})
new_dataframe[new_dataframe.Churn > m]

Churn
State          
AR     0.200000
CA     0.264706
CT     0.162162
DE     0.147541
GA     0.148148
KS     0.185714
MA     0.169231
MD     0.242857
ME     0.209677
MI     0.219178
MN     0.178571
MS     0.215385
MT     0.205882
NC     0.161765
NH     0.160714
NJ     0.264706
NV     0.212121
NY     0.180723
OK     0.147541
PA     0.177778
SC     0.233333
TX     0.250000
WA     0.212121

## Упражнения.

**Упражнение 3.** Найдите среднее количество звонков `Total day calls` для всего датафрейма.

In [100]:
f.groupby(["State"]).agg({"Total day calls": "mean"}).mean()[0]

100.4676583456622

**Упражнение 4.** Найдите среднее количество звонков `Total day calls` для любого выбранного вами штата.

In [105]:
f.groupby(f.State == "OK").agg({"Total day calls": "mean"}).mean()[0]

100.35058519379534

**Упражнение 5.** Создайте датафрейм, в котором будет среднее количество звонков `Total day calls` для каждого штата.

In [107]:
f.groupby(["State"]).agg({"Total day calls": "mean"})

Total day calls
State                 
AK           95.269231
AL           99.100000
AR          102.145455
AZ          100.109375
CA          104.500000
CO           95.121212
CT           98.432432
DC          100.166667
DE           98.639344
FL          101.000000
GA          103.055556
HI          102.811321
IA          103.045455
ID          106.219178
IL          101.689655
IN          107.056338
KS          101.028571
KY          100.338983
LA          103.509804
MA          102.292308
MD           99.357143
ME           99.500000
MI          101.205479
MN           99.119048
MO           99.936508
MS          100.707692
MT           95.220588
NC          102.794118
ND          100.145161
NE           97.409836
NH           98.964286
NJ          100.102941
NM          102.241935
NV          101.424242
NY           98.240964
OH           99.628205
OK          100.262295
OR           95.730769
PA           98.644444
RI          100.430769
SC          102.466667
SD           94.816667
TN          101.452830
TX           99.986111
UT           98.958333
VA          105.363636
VT          100.328767
WA          101.196970
WI           96.679487
WV          103.783019
WY          102.220779

**Упражнение 6.** Оставьте в созданном датафрейме строки только с теми штатами, где количество звонков `Total day calls` больше среднего по исходному датафрейму.

In [109]:
new_dataframe = f.groupby("State").agg({"Total day calls": "mean"})
new_dataframe[new_dataframe["Total day calls"] > f.groupby(["State"]).agg({"Total day calls": "mean"}).mean()[0]]

Total day calls
State                 
AR          102.145455
CA          104.500000
FL          101.000000
GA          103.055556
HI          102.811321
IA          103.045455
ID          106.219178
IL          101.689655
IN          107.056338
KS          101.028571
LA          103.509804
MA          102.292308
MI          101.205479
MS          100.707692
NC          102.794118
NM          102.241935
NV          101.424242
SC          102.466667
TN          101.452830
VA          105.363636
WA          101.196970
WV          103.783019
WY          102.220779

**Упражнение 7.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата.

In [110]:
f.groupby(["State"]).agg({"Total day calls": "mean", "Total eve calls": "mean"})

Total day calls  Total eve calls
State                                  
AK           95.269231        96.576923
AL           99.100000        98.462500
AR          102.145455        97.836364
AZ          100.109375       102.406250
CA          104.500000       103.088235
CO           95.121212        98.166667
CT           98.432432        99.121622
DC          100.166667       101.981481
DE           98.639344        97.360656
FL          101.000000       108.920635
GA          103.055556       103.888889
HI          102.811321        98.849057
IA          103.045455        99.545455
ID          106.219178       100.054795
IL          101.689655       100.810345
IN          107.056338        97.943662
KS          101.028571       101.528571
KY          100.338983        98.237288
LA          103.509804        96.627451
MA          102.292308        98.230769
MD           99.357143        98.571429
ME           99.500000       104.596774
MI          101.205479       101.986301
MN           99.119048       101.642857
MO           99.936508       101.047619
MS          100.707692       100.200000
MT           95.220588        97.514706
NC          102.794118       103.941176
ND          100.145161       105.467742
NE           97.409836       103.590164
NH           98.964286        95.803571
NJ          100.102941       103.441176
NM          102.241935       102.983871
NV          101.424242        96.257576
NY           98.240964       100.024096
OH           99.628205       101.205128
OK          100.262295       104.278689
OR           95.730769        98.500000
PA           98.644444       101.622222
RI          100.430769        94.969231
SC          102.466667        99.783333
SD           94.816667        98.633333
TN          101.452830       103.754717
TX           99.986111       101.902778
UT           98.958333        98.027778
VA          105.363636        99.948052
VT          100.328767        97.671233
WA          101.196970        95.742424
WI           96.679487        99.538462
WV          103.783019        98.056604
WY          102.220779        99.233766

**Упражнение 8.** Создайте датафрейм, в котором будует средние количества звонков `Total day calls` и `Total eve calls` для каждого штата, а также столбец со значениями True и False - ответом на вопрос, больше ли дневных звонков, чем вечерних.

In [124]:
o_df = f.groupby(["State"]).agg({"Total day calls": "mean", "Total eve calls": "mean"})
o_df["Day > eve"] = o_df["Total day calls"] > o_df["Total eve calls"]
o_df

Total day calls  Total eve calls  Day > eve
State                                             
AK           95.269231        96.576923      False
AL           99.100000        98.462500       True
AR          102.145455        97.836364       True
AZ          100.109375       102.406250      False
CA          104.500000       103.088235       True
CO           95.121212        98.166667      False
CT           98.432432        99.121622      False
DC          100.166667       101.981481      False
DE           98.639344        97.360656       True
FL          101.000000       108.920635      False
GA          103.055556       103.888889      False
HI          102.811321        98.849057       True
IA          103.045455        99.545455       True
ID          106.219178       100.054795       True
IL          101.689655       100.810345       True
IN          107.056338        97.943662       True
KS          101.028571       101.528571      False
KY          100.338983        98.237288       True
LA          103.509804        96.627451       True
MA          102.292308        98.230769       True
MD           99.357143        98.571429       True
ME           99.500000       104.596774      False
MI          101.205479       101.986301      False
MN           99.119048       101.642857      False
MO           99.936508       101.047619      False
MS          100.707692       100.200000       True
MT           95.220588        97.514706      False
NC          102.794118       103.941176      False
ND          100.145161       105.467742      False
NE           97.409836       103.590164      False
NH           98.964286        95.803571       True
NJ          100.102941       103.441176      False
NM          102.241935       102.983871      False
NV          101.424242        96.257576       True
NY           98.240964       100.024096      False
OH           99.628205       101.205128      False
OK          100.262295       104.278689      False
OR           95.730769        98.500000      False
PA           98.644444       101.622222      False
RI          100.430769        94.969231       True
SC          102.466667        99.783333       True
SD           94.816667        98.633333      False
TN          101.452830       103.754717      False
TX           99.986111       101.902778      False
UT           98.958333        98.027778       True
VA          105.363636        99.948052       True
VT          100.328767        97.671233       True
WA          101.196970        95.742424       True
WI           96.679487        99.538462      False
WV          103.783019        98.056604       True
WY          102.220779        99.233766       True

**Упражнение 9.** Найти долю клиентов (отношение их числа к общему количеству клиентов) с `international plan` и `voice mail plan`.

In [ ]:
f[(f["International plan"] == "Yes") & (f["Voice mail plan"] == "Yes")].count()[0] / f.count()[0]

0.027602760276027604

**Упражнение 10.** Найти число уникальных значений `Area code`.

In [177]:
f["Area code"].nunique()


3

**Упражнение 11.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; число клиентов, звонивших столько раз. Подсказка: используйте функцию агрегации `count`.

In [ ]:
df11 = f.groupby("Customer service calls").count()
df11 = df11.iloc[:,:1]
df11 = df11.rename( columns = {"State":"Number of customers"})
df11

**Упражнение 12.** Вывести DataFrame из 2 столбцов: число звонков в поддержку; доля оттока (`Churn`). Построить график.

In [ ]:
df12 = f[["Customer service calls", "Churn"]][f.Churn==True].groupby("Customer service calls").count()
df12.Churn = df12.Churn/df11["Number of customers"]*100
df12 = df12.rename( columns = {"Churn":"Churn, %"})
df12
from matplotlib import pyplot as plt
plt.figure(figsize=(10,5))
x = df12.index.values
y = df12.values
plt.scatter(x,y)
plt.minorticks_on()
plt.grid(b=True, which='major', axis='both', alpha=1)
plt.grid(b=True, which='minor', axis='both', alpha=0.5)

**Упражнение 13.** Найти среднюю длительность международного (`intl`) звонка.

In [178]:
f.groupby(["State"]).agg({"Total intl minutes": "mean"}).mean()[0]

10.226460497912981

**Упражнение 14*.** Какие звонки дольше - дневные, вечерние или ночные? Ответ привести в формате DataFrame $3*3$: строки - число минут, число звонков, среднее время звонка. Столбцы - день, вечер, ночь

In [ ]:
tdm_mean = f["Total day minutes"].mean()
tem_mean = f["Total eve minutes"].mean()
tnm_mean = f["Total night minutes"].mean()
tdc_sum = f["Total day calls"].sum()
tec_sum = f["Total eve calls"].sum()
tnc_sum = f["Total night calls"].sum()
tdm_sum = f["Total day minutes"].sum()
tem_sum = f["Total eve minutes"].sum()
tnm_sum = f["Total night minutes"].sum()
data = {'Day' : [tdm_sum, tdc_sum, tdm_mean], 'Eve': [tem_sum, tec_sum, tem_mean], 'Night': [tnm_sum, tnc_sum, tnm_mean]}
df14 = DataFrame(data, index=['Number of minutes', 'Number of calls', 'Mean time of call'])
df14

**Упражнение 15.** Сравнить `Total day charge` для оставшихся и ушедших клиентов.

In [2]:
df15 = f[["Total day calls","Churn"]].groupby("Churn").mean()
df15

NameError: name 'f' is not defined

**Упражнение 16.** Отсортриуйте штаты по `Total day charge` (по возрастанию).

In [188]:
f.groupby(["State"]).agg({"Total day charge": "mean"}).sort_values(by="Total day charge", ascending = True)

Total day charge
State                  
SC            28.295833
RI            28.472154
MO            28.987302
DC            29.135370
NM            29.142903
AZ            29.173125
IL            29.511552
KY            29.539153
WV            29.572358
MT            29.581618
DE            29.680164
NY            29.770361
CT            29.774324
TN            29.881698
HI            29.913019
AR            29.940000
OR            29.962821
NV            29.992879
VA            30.131688
NH            30.146250
NE            30.169344
IA            30.195455
MS            30.248462
LA            30.324706
AK            30.325769
ID            30.365753
CO            30.381818
WA            30.386667
WI            30.452308
FL            30.521429
OK            30.584754
MA            30.617538
WY            30.628961
MI            30.700959
TX            30.858472
VT            30.946301
OH            31.157564
MN            31.170952
CA            31.206176
UT            31.207222
GA            31.455185
NC            31.475294
ME            31.495806
AL            31.621500
ND            31.848548
PA            32.024444
SD            32.247667
KS            32.565143
NJ            33.358676
IN            33.410000
MD            33.529571

**Упражнение 17.** Сделайте агрегацию по средним показателям для каждой `Area code`.

In [ ]:
df17 = f.groupby("Area code").mean()
df17

**Упражнение 18.** Выведите датафрейм размера $3*2$: столбцы State, Churn для 100, 102 и 104 строк исходного датафрейма.

In [ ]:
df18_1 = f[["State", "Churn"]]
df18 = df18_1.iloc[100:105:2]
df18

**Упражнение 19.** Создайте датафрейм из 2 столбцов, заполненных произвольными числами. Добавьте третий столбец, равный их сумме квадратов.

In [1]:
df19 = DataFrame({'A':[1,2,3,4,5], 'B':[30,75,45,7, 1]})
df19["C"] = df19["A"]**2 + df19["B"]**2
df19

NameError: name 'DataFrame' is not defined

**Упражнение 20.** Добавьте к созданному датафрейму четвертый столбец, равный среднему значению первых трех, *используя* функцию `mean`.

In [ ]:
df19["MEAN"] = df19[["A", "B", "C"]].apply(lambda x: x.mean(), axis=1)
df19